### PR TITLE
HDFS-17417. [FGL] HeartbeatManager and DatanodeAdminMonitor support fine-grained locking

### PR DESCRIPTION
**Pull request by ZanderXu**
HeartbeatManager and DatanodeAdminMonitor support fine-grained lock.

`blockManager.processExtraRedundancyBlocksOnInService(node)` in `stopMaintenance` and `stopDecommission` involves storage policy, so it needs FSReadLock.

`isSufficient` in `DatanodeAdminDefaultMonitor` involves UC files, so it needs FSReadLock.

`isBlockReplicatedOk`  in `DatanodeAdminBackoffMonitor` involves UC files, so it needs FSReadLock.

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 52s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  1s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  1s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
|||| _ HDFS-17384 Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  48m 53s |  |  HDFS-17384 passed  |
| +1 :green_heart: |  compile  |   1m 24s |  |  HDFS-17384 passed with JDK Ubuntu-11.0.22+7-post-Ubuntu-0ubuntu220.04.1  |
| +1 :green_heart: |  compile  |   1m 15s |  |  HDFS-17384 passed with JDK Private Build-1.8.0_402-8u402-ga-2ubuntu1~20.04-b06  |
| +1 :green_heart: |  checkstyle  |   1m 15s |  |  HDFS-17384 passed  |
| +1 :green_heart: |  mvnsite  |   1m 26s |  |  HDFS-17384 passed  |
| +1 :green_heart: |  javadoc  |   1m 10s |  |  HDFS-17384 passed with JDK Ubuntu-11.0.22+7-post-Ubuntu-0ubuntu220.04.1  |
| +1 :green_heart: |  javadoc  |   1m 41s |  |  HDFS-17384 passed with JDK Private Build-1.8.0_402-8u402-ga-2ubuntu1~20.04-b06  |
| +1 :green_heart: |  spotbugs  |   3m 21s |  |  HDFS-17384 passed  |
| +1 :green_heart: |  shadedclient  |  40m 50s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   1m 12s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 15s |  |  the patch passed with JDK Ubuntu-11.0.22+7-post-Ubuntu-0ubuntu220.04.1  |
| +1 :green_heart: |  javac  |   1m 15s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 10s |  |  the patch passed with JDK Private Build-1.8.0_402-8u402-ga-2ubuntu1~20.04-b06  |
| +1 :green_heart: |  javac  |   1m 10s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  checkstyle  |   1m  3s |  |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   1m 15s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   0m 57s |  |  the patch passed with JDK Ubuntu-11.0.22+7-post-Ubuntu-0ubuntu220.04.1  |
| +1 :green_heart: |  javadoc  |   1m 30s |  |  the patch passed with JDK Private Build-1.8.0_402-8u402-ga-2ubuntu1~20.04-b06  |
| +1 :green_heart: |  spotbugs  |   3m 21s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  41m  4s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| -1 :x: |  unit  | 266m 43s | [/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6656/1/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt) |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 50s |  |  The patch does not generate ASF License warnings.  |
|  |   | 423m 52s |  |  |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.protocol.TestBlockListAsLongs |
|   | hadoop.hdfs.server.blockmanagement.TestBlockManager |
|   | hadoop.hdfs.tools.TestDFSAdmin |
|   | hadoop.hdfs.server.datanode.TestLargeBlockReport |
|   | hadoop.hdfs.server.blockmanagement.TestDatanodeManager |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.45 ServerAPI=1.45 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6656/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/6656 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets |
| uname | Linux d4f127215181 5.15.0-94-generic #104-Ubuntu SMP Tue Jan 9 15:25:40 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | HDFS-17384 / 1efcabb2097857f08832f5119938ee5053959c50 |
| Default Java | Private Build-1.8.0_402-8u402-ga-2ubuntu1~20.04-b06 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.22+7-post-Ubuntu-0ubuntu220.04.1 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_402-8u402-ga-2ubuntu1~20.04-b06 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6656/1/testReport/ |
| Max. process+thread count | 2536 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs U: hadoop-hdfs-project/hadoop-hdfs |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6656/1/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSNamesystem.java [hunk 1]
```diff
@@ -8713,7 +8714,7 @@ void removeXAttr(String src, XAttr xAttr, boolean logRetryCache)
 
   @Override
   public void removeXattr(long id, String xattrName) throws IOException {
-    writeLock();
+    writeLock(FSNamesystemLockMode.FS);
```
Comment (huangzhaobo99) [inline hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSNamesystem.java]: Hi @ZanderXu sir, removeXAttr RPC seems unrelated to PR title.

**Comment by ferhui** [comment]
failed cases are related to this PR.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   6m 53s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
|||| _ HDFS-17384 Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  32m 41s |  |  HDFS-17384 passed  |
| +1 :green_heart: |  compile  |   0m 43s |  |  HDFS-17384 passed with JDK Ubuntu-11.0.22+7-post-Ubuntu-0ubuntu220.04.1  |
| +1 :green_heart: |  compile  |   0m 41s |  |  HDFS-17384 passed with JDK Private Build-1.8.0_402-8u402-ga-2ubuntu1~20.04-b06  |
| +1 :green_heart: |  checkstyle  |   0m 39s |  |  HDFS-17384 passed  |
| +1 :green_heart: |  mvnsite  |   0m 46s |  |  HDFS-17384 passed  |
| +1 :green_heart: |  javadoc  |   0m 41s |  |  HDFS-17384 passed with JDK Ubuntu-11.0.22+7-post-Ubuntu-0ubuntu220.04.1  |
| +1 :green_heart: |  javadoc  |   1m  6s |  |  HDFS-17384 passed with JDK Private Build-1.8.0_402-8u402-ga-2ubuntu1~20.04-b06  |
| +1 :green_heart: |  spotbugs  |   1m 50s |  |  HDFS-17384 passed  |
| +1 :green_heart: |  shadedclient  |  20m 54s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 37s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 38s |  |  the patch passed with JDK Ubuntu-11.0.22+7-post-Ubuntu-0ubuntu220.04.1  |
| +1 :green_heart: |  javac  |   0m 38s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 34s |  |  the patch passed with JDK Private Build-1.8.0_402-8u402-ga-2ubuntu1~20.04-b06  |
| +1 :green_heart: |  javac  |   0m 34s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  checkstyle  |   0m 31s |  |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   0m 38s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   0m 31s |  |  the patch passed with JDK Ubuntu-11.0.22+7-post-Ubuntu-0ubuntu220.04.1  |
| +1 :green_heart: |  javadoc  |   0m 59s |  |  the patch passed with JDK Private Build-1.8.0_402-8u402-ga-2ubuntu1~20.04-b06  |
| +1 :green_heart: |  spotbugs  |   1m 42s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  20m 50s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| -1 :x: |  unit  | 205m 12s | [/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6656/2/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt) |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 30s |  |  The patch does not generate ASF License warnings.  |
|  |   | 301m  7s |  |  |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.server.datanode.TestLargeBlockReport |
|   | hadoop.hdfs.protocol.TestBlockListAsLongs |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.45 ServerAPI=1.45 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6656/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/6656 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets |
| uname | Linux 4f51d74eac6f 5.15.0-94-generic #104-Ubuntu SMP Tue Jan 9 15:25:40 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | HDFS-17384 / ffa49bdef6465882360b878171158fc38c3158b8 |
| Default Java | Private Build-1.8.0_402-8u402-ga-2ubuntu1~20.04-b06 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.22+7-post-Ubuntu-0ubuntu220.04.1 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_402-8u402-ga-2ubuntu1~20.04-b06 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6656/2/testReport/ |
| Max. process+thread count | 4052 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs U: hadoop-hdfs-project/hadoop-hdfs |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6656/2/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

**Review by huangzhaobo99 — APPROVED** [review]

**Comment by huangzhaobo99** [comment]
LGTM!

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |  18m  4s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  1s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  1s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
|||| _ HDFS-17384 Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  49m 36s |  |  HDFS-17384 passed  |
| +1 :green_heart: |  compile  |   1m 26s |  |  HDFS-17384 passed with JDK Ubuntu-11.0.22+7-post-Ubuntu-0ubuntu220.04.1  |
| +1 :green_heart: |  compile  |   1m 18s |  |  HDFS-17384 passed with JDK Private Build-1.8.0_402-8u402-ga-2ubuntu1~20.04-b06  |
| +1 :green_heart: |  checkstyle  |   1m 15s |  |  HDFS-17384 passed  |
| +1 :green_heart: |  mvnsite  |   1m 28s |  |  HDFS-17384 passed  |
| +1 :green_heart: |  javadoc  |   1m 12s |  |  HDFS-17384 passed with JDK Ubuntu-11.0.22+7-post-Ubuntu-0ubuntu220.04.1  |
| +1 :green_heart: |  javadoc  |   1m 46s |  |  HDFS-17384 passed with JDK Private Build-1.8.0_402-8u402-ga-2ubuntu1~20.04-b06  |
| +1 :green_heart: |  spotbugs  |   3m 24s |  |  HDFS-17384 passed  |
| +1 :green_heart: |  shadedclient  |  40m 46s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   1m 14s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 16s |  |  the patch passed with JDK Ubuntu-11.0.22+7-post-Ubuntu-0ubuntu220.04.1  |
| +1 :green_heart: |  javac  |   1m 16s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   1m  9s |  |  the patch passed with JDK Private Build-1.8.0_402-8u402-ga-2ubuntu1~20.04-b06  |
| +1 :green_heart: |  javac  |   1m  9s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  checkstyle  |   1m  3s |  |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   1m 17s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   0m 58s |  |  the patch passed with JDK Ubuntu-11.0.22+7-post-Ubuntu-0ubuntu220.04.1  |
| +1 :green_heart: |  javadoc  |   1m 39s |  |  the patch passed with JDK Private Build-1.8.0_402-8u402-ga-2ubuntu1~20.04-b06  |
| +1 :green_heart: |  spotbugs  |   3m 23s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  40m 48s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| -1 :x: |  unit  | 267m 19s | [/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6656/3/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt) |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 46s |  |  The patch does not generate ASF License warnings.  |
|  |   | 442m 15s |  |  |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.server.diskbalancer.command.TestDiskBalancerCommand |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.45 ServerAPI=1.45 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6656/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/6656 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets |
| uname | Linux e615cecef0da 5.15.0-94-generic #104-Ubuntu SMP Tue Jan 9 15:25:40 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | HDFS-17384 / f66fbb4bbd87fe682fc21b1a86be4958bee95fa8 |
| Default Java | Private Build-1.8.0_402-8u402-ga-2ubuntu1~20.04-b06 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.22+7-post-Ubuntu-0ubuntu220.04.1 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_402-8u402-ga-2ubuntu1~20.04-b06 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6656/3/testReport/ |
| Max. process+thread count | 2765 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs U: hadoop-hdfs-project/hadoop-hdfs |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6656/3/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by ZanderXu** [comment]
@ferhui The failed UT `hadoop.hdfs.server.diskbalancer.command.TestDiskBalancerCommand` work fine locally.

**Review by ferhui — APPROVED** [review]
+1

**Comment by ferhui** [comment]
@ZanderXu Thanks for contribution. @huangzhaobo99 Thanks for review. Merged.
